keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentConstraintRunner.kt

```diff
@@ -0,0 +1,259 @@
+   */
+  private fun dealWithOldVersions(
+    envInfo: EnvironmentInfo,
+    pendingVersionsToCheck: MutableSet<String>
+  ) {
+    log.debug("pendingVersionsToCheck: [$pendingVersionsToCheck] of artifact ${envInfo.artifact.name} for environment ${envInfo.environment.name} ")
+    pendingVersionsToCheck
+      .sortedWith(envInfo.artifact.versioningStrategy.comparator.reversed()) // oldest first
+      .forEach { version ->
+        val passesConstraints =
+          checkStatelessConstraints(envInfo.artifact, envInfo.deliveryConfig, version, envInfo.environment) &&
+            checkStatefulConstraints(envInfo.artifact, envInfo.deliveryConfig, version, envInfo.environment)
+
+        if (passesConstraints) {
+          repository.queueAllConstraintsApproved(envInfo.deliveryConfig.name, envInfo.environment.name, version)
+        }
+      }
+  }
+
+  /**
+   * A container to hold all info needed for these functions
+   * so that we can actually break this shit up into functions
+   */
+  data class EnvironmentInfo(
+    val deliveryConfig: DeliveryConfig,
+    val environment: Environment,
+    val artifact: DeliveryArtifact,
+    val versions: List<String>,
+    val vetoedArtifacts: Map<String, EnvironmentArtifactVetoes>
+  )
+
+  /**
+   * Queues a version for approval if it's not already approved in the environment.
+   * [EnvironmentPromotionChecker] handles actually approving a version for an environment.
+   */
+  private fun queueApproval(
+    deliveryConfig: DeliveryConfig,
+    artifact: DeliveryArtifact,
+    version: String,
+    targetEnvironment: String
+  ) {
+    val latestVersion = repository
+      .latestVersionApprovedIn(deliveryConfig, artifact, targetEnvironment)
+    if (latestVersion != version) {
+      log.debug("Queueing version $version of ${artifact.type} artifact ${artifact.name} in environment $targetEnvironment for approval")
+      repository.queueAllConstraintsApproved(deliveryConfig.name, targetEnvironment, version)
+    }
+  }
+
+  fun checkStatelessConstraints(
+    artifact: DeliveryArtifact,
+    deliveryConfig: DeliveryConfig,
+    version: String,
+    environment: Environment
+  ): Boolean =
+    checkImplicitConstraints(implicitStatelessEvaluators, artifact, deliveryConfig, version, environment) &&
+      checkConstraints(statelessEvaluators, artifact, deliveryConfig, version, environment)
+
+  fun checkStatefulConstraints(
+    artifact: DeliveryArtifact,
+    deliveryConfig: DeliveryConfig,
+    version: String,
+    environment: Environment
+  ): Boolean =
+    checkImplicitConstraints(implicitStatefulEvaluators, artifact, deliveryConfig, version, environment) &&
+      checkConstraints(statefulEvaluators, artifact, deliveryConfig, version, environment)
+
+  /**
+   * Checks constraints for a list of evaluators.
+   * Evaluates the constraint for every environment passed in.
+   * @return true if all constraints pass
+   */
+  private fun checkImplicitConstraints(
```

Comment:
I'm going to leave it as is. It's a confusing piece of code, and the `implicit` in the function name refers to the way in which we check constraints (check it for every environment, because it won't be defined on the environment)

Reply:
Right, but to me that's exactly what was weird: this function calls all the evaluators passed in regardless of the environment, as opposed to the one down below which only runs supported evaluators for the environment. A function name should describe what it does, not what its intended usage is (which is entirely contextual), IMHO. There's nothing preventing anyone from calling this function with the wrong list of evaluators -- it would blindly run all them without checking that the constraints are implicit. My 2 cents. 🤷‍♂️

Reply:
ah okay, I see what you're saying. Trying to keep all the old + new in my head is hard :D I'll rename for function instead of desired usage.